vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs

```diff
@@ -30,83 +34,292 @@ type internal FSharpNavigableItem(document: Document, textSpan: TextSpan) =
+                    | true , None      , Some range, None       -> return Some range 
+                    | false, _         , Some range, _          -> return Some range 
+                    | false, _         , None      , Some range -> return Some range 
+                    | false, Some range, None      , None       -> return Some range 
+                    | _    , None      , None      , None       -> 
+                        let! findSigDeclarationResult = 
+                            checkFileResults.GetDeclarationLocationAlternate
+                                (idRange.StartLine, idRange.EndColumn, lineText, lexerSymbol.FullIsland, preferFlag=preferSignature)
+                
+                        match findSigDeclarationResult with 
+                        | FSharpFindDeclResult.DeclNotFound _failReason -> return None 
+                        | FSharpFindDeclResult.DeclFound declRange -> return Some declRange
+                }
+                let! targetDocument = originDocument.Project.Solution.TryGetDocumentFromFSharpRange (targetRange, originDocument.Project.Id)
+                return! rangeToNavigableItem (targetRange, targetDocument)
+        }  
+
+
+    /// find the declaration location (signature file/.fsi) of the target symbol if possible, fall back to definition 
+    let findDeclarationOfSymbolAtRange
+        (document:Document, range:range, sourceText:SourceText, checker: FSharpChecker, projectInfoManager: ProjectInfoManager) =
+        findSymbolHelper (document, range, sourceText,true, checker, projectInfoManager) 
+
+
+    /// find the definition location (implementation file/.fs) of the target symbol
+    let findDefinitionOfSymbolAtRange
+        (document:Document, range:range, sourceText:SourceText, checker: FSharpChecker, projectInfoManager: ProjectInfoManager) =
+        findSymbolHelper (document, range, sourceText,false, checker, projectInfoManager)
+
+
+
+[<Shared>]
+[<ExportLanguageService(typeof<IGoToDefinitionService>, FSharpCommonConstants.FSharpLanguageName)>]
+[<Export(typeof<FSharpGoToDefinitionService>)>]
+type internal FSharpGoToDefinitionService [<ImportingConstructor>]
+    (checkerProvider: FSharpCheckerProvider,
+     projectInfoManager: ProjectInfoManager,
+     [<ImportMany>] presenters: IEnumerable<INavigableItemsPresenter>) =
+
+    let serviceProvider =  ServiceProvider.GlobalProvider  
+    let statusBar = serviceProvider.GetService<SVsStatusbar,IVsStatusbar>()
+
+    /// helper composition function to construct a task from a function that returns an Async Computation and its arguments
+    let createNavigationTask (navFunction: _ -> Async<FSharpNavigableItem  option>) args (cancellationToken:CancellationToken) = 
+        asyncMaybe {
+            try let navList = List<INavigableItem>()
+                let! navItem = navFunction args 
+                navList.Add navItem 
+                return navList.AsEnumerable()
```

Comment:
Returning `seq` from an async computation looks a bit unsafe. In this case it's ok, but I'd prefer an array or a list.